# TRANSACTIONS DATASET EXPLORATION USING PYTHON

# 💳 Transactions Data Validation

This notebook focuses on validating the transactions dataset used in the Coffee Shop BI project.

The objective is to ensure data integrity, consistency, and readiness for further processing in Power BI.

#### Installation of pandas and import pandas and os 

In [1]:
!pip install pandas
import os
import pandas as pd


[notice] A new release of pip is available: 26.0.1 -> 26.1
[notice] To update, run: C:\Users\keren mayer\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable


## 1. Data Loading

#### Inspecting available sheets in the Excel file

In [2]:
file_path = "OneDrive/Documents/Bureau/EXPERIS/FINAL BI PROJECT/data/Coffee Shop Sales (2)good.xlsx"

xls = pd.ExcelFile(file_path)
xls.sheet_names

['Pivot Tables', 'Dashboard', 'Transactions']

#### Loading a specific sheet from the Excel file

In [3]:
df = pd.read_excel(file_path, sheet_name="Transactions")
df.head()

,transaction_id,transaction_date,transaction_time,transaction_qty,store_id,store_location,product_id,unit_price,product_category,product_type,product_detail,Revenue,Month,Month.1,Weekday,Weekday.1,Hour
0,1,2023-01-01,07:06:11,2,5,Lower Manhattan,32,3.0,Coffee,Gourmet brewed coffee,Ethiopia Rg,6.0,1,Jan,7,Sun,7
1,2,2023-01-01,07:08:56,2,5,Lower Manhattan,57,3.1,Tea,Brewed Chai tea,Spicy Eye Opener Chai Lg,6.2,1,Jan,7,Sun,7
2,3,2023-01-01,07:14:04,2,5,Lower Manhattan,59,4.5,Drinking Chocolate,Hot chocolate,Dark chocolate Lg,9.0,1,Jan,7,Sun,7
3,4,2023-01-01,07:20:24,1,5,Lower Manhattan,22,2.0,Coffee,Drip coffee,Our Old Time Diner Blend Sm,2.0,1,Jan,7,Sun,7
4,5,2023-01-01,07:22:41,2,5,Lower Manhattan,57,3.1,Tea,Brewed Chai tea,Spicy Eye Opener Chai Lg,6.2,1,Jan,7,Sun,7


## 2. Initial Data Exploration

#### Inspecting the overall structure of the dataset

In [15]:
df.shape

(149116, 17)

#### Reviewing column names and dataset schema

In [16]:
df.columns

Index(['transaction_id', 'transaction_date', 'transaction_time',
       'transaction_qty', 'store_id', 'store_location', 'product_id',
       'unit_price', 'product_category', 'product_type', 'product_detail',
       'Revenue', 'Month', 'Month.1', 'Weekday', 'Weekday.1', 'Hour'],
      dtype='object')

## 3. Data Structure Analysis

#### Checking data types and completeness of the dataset

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 149116 entries, 0 to 149115
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   transaction_id    149116 non-null  int64         
 1   transaction_date  149116 non-null  datetime64[ns]
 2   transaction_time  149116 non-null  object        
 3   transaction_qty   149116 non-null  int64         
 4   store_id          149116 non-null  int64         
 5   store_location    149116 non-null  object        
 6   product_id        149116 non-null  int64         
 7   unit_price        149116 non-null  float64       
 8   product_category  149116 non-null  object        
 9   product_type      149116 non-null  object        
 10  product_detail    149116 non-null  object        
 11  Revenue           149116 non-null  float64       
 12  Month             149116 non-null  int64         
 13  Month.1           149116 non-null  object        
 14  Week

## 4. Missing Values Analysis

#### Assessing missing values across all columns

In [9]:
df.isnull().sum()

transaction_id      0
transaction_date    0
transaction_time    0
transaction_qty     0
store_id            0
store_location      0
product_id          0
unit_price          0
product_category    0
product_type        0
product_detail      0
Revenue             0
Month               0
Month.1             0
Weekday             0
Weekday.1           0
Hour                0
dtype: int64

## 5. Duplicate Records Check

#### Checking for duplicate transactions

In [10]:
df.duplicated().sum()

np.int64(0)

No duplicate transactions detected.

## 6. Revenue Consistency Check

#### Validating revenue consistency

In [11]:
(df["transaction_qty"] * df["unit_price"] - df["Revenue"]).abs().describe()

count    149116.0
mean          0.0
std           0.0
min           0.0
25%           0.0
50%           0.0
75%           0.0
max           0.0
dtype: float64

Revenue calculation is consistent across the dataset.

## 6. Statistical Summary of Key Metrics

This section reviews the distribution of key numerical variables:
- transaction quantity
- unit price
- revenue

The goal is to identify potential anomalies or inconsistencies in the data.

In [12]:
# decsribe generate decriptive statistics
df[["transaction_qty", "unit_price", "Revenue"]].describe()

,transaction_qty,unit_price,Revenue
count,149116.000000,149116.000000,149116.000000
mean,1.438276,3.382219,4.686367
std,0.542509,2.658723,4.227099
min,1.000000,0.800000,0.800000
25%,1.000000,2.500000,3.000000
50%,1.000000,3.000000,3.750000
75%,2.000000,3.750000,6.000000
max,8.000000,45.000000,360.000000


### Key Observations

- Transaction quantity is generally low (median = 1), indicating most purchases are single items  
- Unit prices vary significantly, suggesting a mix of low-cost and premium products  
- Revenue distribution is consistent with quantity × price logic  
- Maximum values appear reasonable (no extreme outliers detected)  

Overall, the dataset shows consistent and realistic value ranges.



In [17]:
df.sample(17)

,transaction_id,transaction_date,transaction_time,transaction_qty,store_id,store_location,product_id,unit_price,product_category,product_type,product_detail,Revenue,Month,Month.1,Weekday,Weekday.1,Hour
104055,104309,2023-05-22,14:55:40,2,3,Astoria,61,4.75,Drinking Chocolate,Hot chocolate,Sustainably Grown Organic Lg,9.50,5,May,1,Mon,14
54750,54861,2023-03-31,14:18:19,1,3,Astoria,71,3.75,Bakery,Pastry,Chocolate Croissant,3.75,3,Mar,5,Fri,14
63110,63280,2023-04-10,18:20:59,1,5,Lower Manhattan,63,0.80,Flavours,Regular syrup,Carmel syrup,0.80,4,Apr,1,Mon,18
45187,45294,2023-03-17,12:03:34,1,8,Hell's Kitchen,37,3.00,Coffee,Barista Espresso,Espresso shot,3.00,3,Mar,5,Fri,12
65470,65640,2023-04-13,12:19:33,1,3,Astoria,26,3.00,Coffee,Organic brewed coffee,Brazilian Rg,3.00,4,Apr,4,Thu,12
21052,21109,2023-02-07,16:17:15,1,3,Astoria,54,2.50,Tea,Brewed Chai tea,Morning Sunrise Chai Rg,2.50,2,Feb,2,Tue,16
61357,61527,2023-04-08,19:24:28,2,3,Astoria,74,3.50,Bakery,Biscotti,Ginger Biscotti,7.00,4,Apr,6,Sat,19
105132,105386,2023-05-23,14:43:50,2,3,Astoria,46,2.50,Tea,Brewed Green tea,Serenity Green Tea Rg,5.00,5,May,2,Tue,14
137733,138070,2023-06-21,08:15:56,1,3,Astoria,69,3.25,Bakery,Biscotti,Hazelnut Biscotti,3.25,6,Jun,3,Wed,8
126157,126494,2023-06-11,13:45:06,1,5,Lower Manhattan,57,3.10,Tea,Brewed Chai tea,Spicy Eye Opener Chai Lg,3.10,6,Jun,7,Sun,13


## 8. Conclusion

- Dataset contains 149116 transactions across multiple locations  
- No critical missing values detected  
- No duplicate records found  
- Revenue calculations are consistent  
- Data is clean and ready for BI analysis  